wsd/DocumentBroker.cpp

```diff
@@ -766,6 +781,188 @@ void DocumentBroker::stop(const std::string& reason)
     _poll->wakeup();
 }
 
+bool DocumentBroker::downloadAdvance(const std::string& jailId,
+                                     const WopiStorage::WOPIFileInfo& wopiFileInfo)
+{
+    ASSERT_CORRECT_THREAD();
+
+    LOG_INF("Loading [" << _docKey << "] ahead-of-time in jail [" << jailId << ']');
+
+    assert(!_docState.isMarkedToDestroy() && "MarkedToDestroy while downloading ahead-of-time");
+
+    _jailId = jailId;
+
+    // The URL is the publicly visible one, not visible in the chroot jail.
+    // We need to map it to a jailed path and copy the file there.
+
+    // /tmp/user/docs/<dirName>, root under getJailRoot()
+    const Poco::Path jailPath(JAILED_DOCUMENT_ROOT, Util::rng::getFilename(16));
+    const std::string jailRoot = getJailRoot();
+
+    LOG_INF("JailPath for docKey [" << _docKey << "]: [" << jailPath.toString() << "], jailRoot: ["
+                                    << jailRoot << ']');
+
+    assert(_storage == nullptr &&
+           "Unexpected to find storage created while downloading ahead-of-time");
+
+    _docState.setStatus(DocumentState::Status::Downloading);
+
+    // Pass the public URI to storage as it needs to load using the token
+    // and other storage-specific data provided in the URI.
+    LOG_DBG("Creating new storage instance for URI ["
+            << COOLWSD::anonymizeUrl(_uriPublic.toString()) << ']');
+
+    try
+    {
+        _storage = StorageBase::create(_uriPublic, jailRoot, jailPath.toString(),
+                                       /*takeOwnership=*/isConvertTo());
+    }
+    catch (...)
+    {
+        // fallthrough
+    }
+
+    if (_storage == nullptr)
+    {
+        // We should get an exception, not null.
+        LOG_WRN("Failed to create storage ahead-of-time for ["
+                << _docKey << "] in [" << jailPath.toString()
+                << "], will try again on user connection");
+        return false;
+    }
+
+    LOG_ASSERT(_storage);
+
+    // Call the storage specific fileinfo functions
+    std::string userId, username;
+    std::string userExtraInfo;
+    std::string userPrivateInfo;
+    std::string watermarkText;
+    std::string templateSource;
+
+#if !MOBILEAPP
+    std::chrono::milliseconds checkFileInfoCallDurationMs = std::chrono::milliseconds::zero();
+    WopiStorage* wopiStorage = dynamic_cast<WopiStorage*>(_storage.get());
```

Comment:
Not a huge fan of this. Not a huge fan of LocalStorage either but I imagine we need the efficiency for mobile (?) - or could we abstract local files helpfully too behind the same server API ?

Reply:
I refrained from making unrelated changes. Agreed this needs clean up. Not a fan either.

LocalStorage is needed many of the tests. Internally, we serve the files through an internal Wopi server, so as to exercise the same code.